Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm

```diff
@@ -6503,13 +6502,75 @@ - (void)_handleKeyUIEvent:(::UIEvent *)event
             _seenHardwareKeyDownInNonEditableElement = YES;
             [self reloadInputViews];
         }
-        [super _handleKeyUIEvent:event];
+        [super pressesBegan:presses withEvent:event];
         return;
     }
 
-    [super _handleKeyUIEvent:event];
+    [super pressesBegan:presses withEvent:event];
+}
+
+- (void)pressesEnded:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    bool isHardwareKeyboardEvent = !!event._hidEvent;
+    // We only want to handle key event from the hardware keyboard when we are
+    // first responder and we are not interacting with editable content.
+    if ([self isFirstResponder] && isHardwareKeyboardEvent && (_inputPeripheral || !_page->editorState().isContentEditable)) {
+        if ([_inputPeripheral respondsToSelector:@selector(handleKeyEvent:)]) {
+            if ([_inputPeripheral handleKeyEvent:event])
+                return;
+        }
+        if (!_seenHardwareKeyDownInNonEditableElement) {
+            _seenHardwareKeyDownInNonEditableElement = YES;
+            [self reloadInputViews];
+        }
+        [super pressesEnded:presses withEvent:event];
+        return;
+    }
+
+    [super pressesEnded:presses withEvent:event];
+}
+
+- (void)pressesChanged:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    bool isHardwareKeyboardEvent = !!event._hidEvent;
+    // We only want to handle key event from the hardware keyboard when we are
+    // first responder and we are not interacting with editable content.
+    if ([self isFirstResponder] && isHardwareKeyboardEvent && (_inputPeripheral || !_page->editorState().isContentEditable)) {
+        if ([_inputPeripheral respondsToSelector:@selector(handleKeyEvent:)]) {
+            if ([_inputPeripheral handleKeyEvent:event])
+                return;
+        }
+        if (!_seenHardwareKeyDownInNonEditableElement) {
+            _seenHardwareKeyDownInNonEditableElement = YES;
+            [self reloadInputViews];
+        }
+        [super pressesChanged:presses withEvent:event];
+        return;
+    }
+
+    [super pressesChanged:presses withEvent:event];
+}
+
+- (void)pressesCancelled:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    bool isHardwareKeyboardEvent = !!event._hidEvent;
+    // We only want to handle key event from the hardware keyboard when we are
+    // first responder and we are not interacting with editable content.
+    if ([self isFirstResponder] && isHardwareKeyboardEvent && (_inputPeripheral || !_page->editorState().isContentEditable)) {
```

Comment:
Hm...it also seems a tiny bit weird to run this logic in the case of `-pressesCancelled`, but I suppose that matches shipping behavior since `-_handleKeyUIEvent:` can be called for cancelled key events, so it's probably best to leave it here.

Reply:
Also made a helper method in `WKContentViewInteraction` to share the logic

Reply:
`- (BOOL)_tryToHandlePressesEvent:(UIPressesEvent *)event`

Reply:
Yeah I'm not quite sure what it _means_ to exercise this logic on a `pressesCancelled` but since this is shipping behavior I didn't probe too much here